File: Problem-lists/Blind75/sliding-window/3_Longest_Substring_Without_Repeating_Characters.cpp
#include <algorithm>
#include <iostream>
#include <unordered_set>
#include <vector>
using namespace std;

/*
    Time: O(N) [Hash map count/insert/erase are O(1) average, Space: O(N) [worst case all character are unqiue]
*/

int lengthOfLongestSubstring(string s) {
    unordered_set<int> seq;
    int start = 0;
    int max_len = 0;
    int cur_len = 0;
    for (auto n : s) {
        if (seq.count(n) == 0) {  // not duplicate character
            seq.insert(n);
            cur_len++;
        } else {
            while (s[start] != n) {
                seq.erase(s[start]);
                cur_len--;
                start++;
            }
            start++;
        }
        max_len = max(max_len, cur_len);
    }

    return max_len;
}

File: Problem-lists/Blind75/sliding-window/3_Longest_Substring_Without_Repeating_Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let start = 0
    var seq = new Set()
    let cur_len = 0
    let max_len = 0
    for (let n of s) {
        if (!seq.has(n)) { // not duplicate
            seq.add(n)
            cur_len++;
        } else {
            while (s[start] != n) {
                seq.delete(s[start]);
                start++;
                cur_len--;
            }
            start++;
        }
        max_len = Math.max(max_len, cur_len);
    }

    return max_len;
};